sycl/doc/extensions/proposed/sycl_ext_oneapi_launch_queries.asciidoc

```diff
@@ -0,0 +1,259 @@
+  sycl_ext_oneapi_kernel_properties]
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+This extension introduces a new mechanism for querying any limitations that a
+developer must respect when launching a specific kernel on a specific queue.
+Such limitations may exist when a kernel is decorated with one or more
+properties that require an implementation to enable additional features
+(such as providing certain forward progress guarantees, or enabling
+cross-work-group synchronization routines), or when a kernel uses certain
+features (such as static work-group local memory or group algorithms).
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_LAUNCH_QUERIES` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Launch queries
+
+An implementation's ability to satisify requests for specific behaviors
+(such as strong forward progress guarantees and/or cross-work-group
+synchronization) for a kernel may be dependent upon: the kernel itself, and the
+features that it uses; the queue (and underlying device) to which the kernel is
+submitted; and the kernel's launch configuration.
+
+It is a user's responsibility to ensure that a kernel requesting specific
+behaviors uses a compatible launch configuration, using the
+`ext_oneapi_get_info` function and descriptors from the `info::kernel`
+namespace.
+
+If the `sycl::nd_range` parameter used to launch a kernel is incompatible with
+the results of a kernel's launch queries, an implementation must throw a
+synchronous exception with the `errc::nd_range` error code.
+
+[NOTE]
+====
+The values returned by `ext_oneapi_get_info` account for all properties
+attached to a kernel (via the mechanisms defined in the
+sycl_ext_oneapi_kernel_properties extension), as well as the usage of features
+like group algorithms and work-group local memory. Developers should assume
+that the values will differ across kernels.
+====
+
+[source,c++]
+----
+namespace sycl {
+
+class kernel {
+ public:
+  template <typename Param, typename... T>
```

Comment:
Would this replace the existing `kernel::get_info()`?

Reply:
Yes, exactly.  To keep this extension simple, I think we'd only implement the new descriptors for now, but my hope is that we could add this in a future version of SYCL and carry the old information descriptors across.